src/main/scala/treadle/executable/Memory.scala

```diff
@@ -590,6 +591,52 @@ object Memory {
       )
     }
   }
+}
+
+class MemoryInitializer(engine: ExecutionEngine) {
+  case class MemoryMetadata(symbol: Symbol, fileName: String, radix: Int)
+
+  val memoryLoadAnnotations: Seq[LoadMemoryAnnotation] = engine.optionsManager.firrtlOptions.annotations.collect {
+    case mla: LoadMemoryAnnotation => mla
+  }
 
+  /*
+  module containing a memory may have been instantiated multiple times
+  This makes sure that every instance of the memory gets loaded
+   */
+  val memoryMetadata: Seq[MemoryMetadata] = memoryLoadAnnotations.flatMap { anno =>
+    anno.target match {
+      case ComponentName(memoryName, ModuleName(moduleName, _)) =>
+        val radix = anno.hexOrBinary match {
+          case MemoryLoadFileType.Hex => 16
+          case MemoryLoadFileType.Binary => 2
+        }
+        engine.symbolTable.moduleMemoryToMemorySymbol(s"$moduleName.$memoryName").toSeq.map { memorySymbol =>
+          MemoryMetadata(memorySymbol, anno.getFileName, radix)
+        }
+      case _ => Seq()
+    }
+  }
 
+  private def doInitialize(memorySymbol: Symbol, fileName: String, radix: Int): Unit = {
+    io.Source.fromFile(fileName).getLines().zipWithIndex.foreach {
+      case (line, lineNumber) if lineNumber < memorySymbol.slots =>
+        try {
+          val value = BigInt(line.trim, radix)
+          engine.dataStore.update(memorySymbol, lineNumber, value)
```

Comment:
Could we encode the radix with the data representation itself? This seems to me to be more a property of the string representing the value than a property of the memory initializer. I would think to convert between base 10, 16, or 2 representations would just require changing the source file containing those representations. Or maybe I misunderstand `anno.hexOrBinary`.

Reply:
There are only two functions in verilog that do the file loading to memory. One uses hex-text and the other uses binary-text.  I think this is a familiar choice they are making and it shouldn't be encoded in the file name

Reply:
@ucbjrl ping